File: Core/graphics/material/Material.cs
using Core.graphics.shader;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace Core.graphics.material
{
    [JsonPolymorphic(TypeDiscriminatorPropertyName = "$type")]
    [JsonDerivedType(typeof(MaterialDefault), "default")]
    [JsonDerivedType(typeof(MaterialPhong), "phong")]
    [JsonDerivedType(typeof(MaterialPBR), "pbr")]
    public abstract class MaterialBase
    {
        public string Id { get; set; } = "";
        public string Name { get; set; } = "";
        public string Path { get; set; } = "";
        public abstract ShaderTypes Shader { get; }
    }
    public class MaterialPhong : MaterialBase
    {
        public override ShaderTypes Shader => ShaderTypes.Phong;
        public ColorOrTexture Ambient { get; set; } = new() { Color = "#ffffff" };
        public ColorOrTexture Diffuse { get; set; } = new() { Color = "#ffffff" };
        public ColorOrTexture Specular { get; set; } = new() { Color = "#ffffff" };
        public float Shininess { get; set; } = 32.0f;
    }

    public class MaterialPBR : MaterialBase
    {
        public override ShaderTypes Shader => ShaderTypes.PBR;

        public string AlbedoMap { get; set; } = "";
        public float Metallic { get; set; } = 0.5f;
        public float Roughness { get; set; } = 0.5f;
        public Vector3 Emissive { get; set; } = new(0, 0, 0);
    }

    public class MaterialDefault : MaterialBase
    {
        public override ShaderTypes Shader => ShaderTypes.Basic;
        public Vector3 DiffuseColor { get; set; } = new(1, 1, 1);
        public Vector3 SpecularColor { get; set; } = new(1, 1, 1);
        public float Shininess { get; set; } = 32.0f;
    }


    public class ColorOrTexture
    {
        public string? Color { get; set; }
        public string? TexturePath { get; set; }
        public bool IsTexture { get; set; } = false;

    }

}

File: EditorC/Material.cpp
#include "Material.h"


std::shared_ptr<Material::MaterialBase> Material::MaterialBase::fromPtree(const boost::property_tree::ptree& pt) {
	std::string type = pt.get<std::string>("$type");

	std::shared_ptr<Material::MaterialBase> mat;

	if (type == "phong") {
		auto phong = std::make_shared<Material::MaterialPhong>();
		phong->shininess = pt.get<float>("Shininess", 32.0f);
		phong->ambient = ColorOrTexture::fromPtree(pt.get_child("Ambient"));
		phong->diffuse = ColorOrTexture::fromPtree(pt.get_child("Diffuse"));
		phong->specular = ColorOrTexture::fromPtree(pt.get_child("Specular"));
		mat = phong;
	}
	else if (type == "pbr") {
		auto pbr = std::make_shared<Material::MaterialPBR>();
		pbr->albedoMap = pt.get<std::string>("albedoMap", "");
		pbr->metallic = pt.get<float>("metallic", 0.5f);
		pbr->roughness = pt.get<float>("roughness", 0.5f);
		pbr->emissive[0] = pt.get<float>("emissive.x", 0);
		pbr->emissive[1] = pt.get<float>("emissive.y", 0);
		pbr->emissive[2] = pt.get<float>("emissive.z", 0);
		mat = pbr;
	}
	else if (type == "default") {
		auto def = std::make_shared<Material::MaterialDefault>();
		def->shininess = pt.get<float>("shininess", 32.0f);
		def->diffuseColor[0] = pt.get<float>("diffuseColor.x", 1);
		def->diffuseColor[1] = pt.get<float>("diffuseColor.y", 1);
		def->diffuseColor[2] = pt.get<float>("diffuseColor.z", 1);
		def->specularColor[0] = pt.get<float>("specularColor.x", 1);
		def->specularColor[1] = pt.get<float>("specularColor.y", 1);
		def->specularColor[2] = pt.get<float>("specularColor.z", 1);
		mat = def;
	}
	else {
		throw std::runtime_error("Unknown material type: " + type);
	}

	// Setează câmpurile comune
	mat->id = pt.get<std::string>("Id", "");
	mat->name = pt.get<std::string>("Name", "");
	mat->path = pt.get<std::string>("Path", "");

	return mat;
}
